File: src/test/java/io/github/marmer/GreetingProviderTest.java
package io.github.marmer;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;

import java.time.LocalTime;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.spy;

class GreetingProviderTest {
    private SystemTimeProvider systemTimeProvider = spy(SystemTimeProvider.class);

    private GreetingProvider underTest = new GreetingProvider(systemTimeProvider);


    private static Object[][] testData() {
        return new Object[][]{
                {LocalTime.of(6, 0), "John Doe", "Buenos Dias John Doe"},
                {LocalTime.of(11, 59, 59, 99999), "John Doe", "Buenos Dias John Doe"},
                {LocalTime.of(7, 2, 3, 45678), "John Doe", "Buenos Dias John Doe"},
                {LocalTime.of(12, 0), "John Doe", "Buenos Tarde John Doe"},
                {LocalTime.of(19, 59, 59, 99999), "John Doe", "Buenos Tarde John Doe"},
                {LocalTime.of(13, 2, 3, 45678), "John Doe", "Buenos Tarde John Doe"},
                {LocalTime.of(20, 0), "John Doe", "Buenos Noche John Doe"},
                {LocalTime.of(5, 59, 59, 99999), "John Doe", "Buenos Noche John Doe"},
                {LocalTime.of(0, 0), "John Doe", "Buenos Noche John Doe"},
                {LocalTime.of(7, 2, 3, 45678), "Superman", "Buenos Dias Superman"},
        };
    }

    @ParameterizedTest
    @MethodSource("testData")
    @DisplayName("It should greet correctly with the given name at the corret time")
    void getGreetingFor_ItShouldGreetCorrectlyWithTheGivenNameAtTheCorretTime(final LocalTime time, final String name, final String expectedGreeting)
            throws Exception {
        // Preparation
        doReturn(time).when(systemTimeProvider).getNow();

        // Execution
        final var result = underTest.getGreetingFor(name);

        // Assertion
        assertEquals(expectedGreeting, result);
    }
}

File: src/test/kotlin/io/github/marmer/GreetingProviderTest.kt
package io.github.marmer

import com.nhaarman.mockitokotlin2.doReturn
import com.nhaarman.mockitokotlin2.spy
import org.junit.jupiter.api.Test
import java.time.LocalTime
import kotlin.test.assertEquals

internal class GreetingProviderTest {
    private var systemTimeProviderMock = spy<SystemTimeProvider>()

    private var underTest: GreetingProvider = GreetingProvider(systemTimeProviderMock)

    @Test
    @Throws(Exception::class)
    fun `The answer should contain buenos dias at beginning of the morning with a name`() {
        // Preparation
        doReturn(LocalTime.of(6, 0)).`when`(systemTimeProviderMock).currentTime()

        // Execution
        val result = underTest.greetingFor("Bart")

        // Assertion
        assertEquals(result, "Buenos dias Bart")
    }

    @Test
    @Throws(Exception::class)
    fun `The answer should contain buenos dias at the ending of morning with a name`() {
        // Preparation
        doReturn(LocalTime.of(11, 59, 59, 99999)).`when`(systemTimeProviderMock).currentTime()

        // Execution
        val result = underTest.greetingFor("Bart")

        // Assertion
        assertEquals(result, "Buenos dias Bart")
    }

    @Test
    @Throws(Exception::class)
    fun `The answer should contain buenos tarde at beginning of the evening with a name`() {
        // Preparation
        doReturn(LocalTime.of(12, 0)).`when`(systemTimeProviderMock).currentTime()

        // Execution
        val result = underTest.greetingFor("Bart")

        // Assertion
        assertEquals(result, "Buenos tarde Bart")
    }

    @Test
    @Throws(Exception::class)
    fun `The answer should contain buenos tarde at the ending of evening with a name`() {
        // Preparation
        doReturn(LocalTime.of(19, 59, 59, 99999)).`when`(systemTimeProviderMock).currentTime()

        // Execution
        val result = underTest.greetingFor("Bart")

        // Assertion
        assertEquals(result, "Buenos tarde Bart")
    }

    @Test
    @Throws(Exception::class)
    fun `The answer should contain buenos noche at beginning of the night with a name`() {
        // Preparation
        doReturn(LocalTime.of(20, 0)).`when`(systemTimeProviderMock).currentTime()

        // Execution
        val result = underTest.greetingFor("Bart")

        // Assertion
        assertEquals(result, "Buenos noche Bart")
    }

    @Test
    @Throws(Exception::class)
    fun `The answer should contain buenos noche at the ending of night with a name`() {
        // Preparation
        doReturn(LocalTime.of(5, 59, 59, 99999)).`when`(systemTimeProviderMock).currentTime()

        // Execution
        val result = underTest.greetingFor("Bart")

        // Assertion
        assertEquals(result, "Buenos noche Bart")
    }

}
